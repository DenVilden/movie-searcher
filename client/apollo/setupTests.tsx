import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { ThemeProvider } from 'styled-components';
import { theme } from '../pages/_app';

// eslint-disable-next-line jest/prefer-spy-on
HTMLElement.prototype.scrollIntoView = jest.fn();

type RenderApolloOptions = {
  mocks?: MockedResponse[];
  addTypeName?: boolean;
};

export const renderApollo = (
  element: React.ReactElement,
  { mocks = [], addTypeName = false }: RenderApolloOptions = {},
) =>
  render(
    <MockedProvider addTypename={addTypeName} mocks={mocks}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MockedProvider>,
  );

export * from '@testing-library/react';
