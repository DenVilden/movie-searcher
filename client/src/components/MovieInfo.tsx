import React from 'react';
import {
  Typography,
  Divider,
  Button,
  CardContent,
  Card,
  CardMedia,
} from '@material-ui/core';
import Head from 'next/head';
import styled from 'styled-components';
import { MovieInfo as MovieInfoType } from '../generated/types';

const StyledCard = styled(Card)`
  background-color: inherit;
  display: block;
  margin: ${(props) => props.theme.spacing(2)}px;

  ${(props) => props.theme.breakpoints.up('sm')} {
    display: flex;
  }
`;

const StyledCardMedia = styled(CardMedia)`
  height: 450px;
  margin: ${(props) => props.theme.spacing(2)}px;

  ${(props) => props.theme.breakpoints.up('sm')} {
    width: 40%;
  }
`;

const StyledCardContent = styled(CardContent)`
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  ${(props) => props.theme.breakpoints.up('sm')} {
    width: 60%;
  }
`;

const StyledTypography = styled(Typography)`
  display: flex;
  justify-content: space-between;
`;

type Props = {
  movie: MovieInfoType;
  loading: boolean;
  addOrRemoveFromFavorites: () => void;
};

const MovieInfo = ({ movie, addOrRemoveFromFavorites, loading }: Props) => (
  <StyledCard elevation={10}>
    <Head>
      <title>{movie.title}</title>
    </Head>
    <StyledCardMedia image={movie.backdrop_path || 'no-image.jpg'} src="img" />
    <StyledCardContent>
      <StyledTypography gutterBottom variant="h5">
        {movie.title}
        <Button
          color={movie.isInFavorites ? 'secondary' : 'primary'}
          data-testid="favorites-button"
          disabled={loading}
          onClick={addOrRemoveFromFavorites}
          variant="contained"
        >
          {/* eslint-disable-next-line no-nested-ternary */}
          {loading
            ? 'Updating'
            : movie.isInFavorites
            ? 'Remove from favorites'
            : 'Add to favorites'}
        </Button>
      </StyledTypography>
      <Typography>{movie.overview}</Typography>
      <Divider />
      <Typography>
        <b>Budget:</b> {movie.budget}
      </Typography>
      <Divider />
      <Typography>
        <b>Revenue:</b> {movie.revenue}
      </Typography>
      <Divider />
      <Typography>
        <b>Rating:</b> {movie.vote_average}
      </Typography>
      <Divider />
      <Typography>
        <b>Release Date:</b> {movie.release_date}
      </Typography>
    </StyledCardContent>
  </StyledCard>
);

export default MovieInfo;
