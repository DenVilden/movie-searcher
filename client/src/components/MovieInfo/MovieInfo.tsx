import React from 'react';
import { Typography, Divider, Button } from '@material-ui/core';
import {
  StyledCard,
  StyledCardMedia,
  StyledCardContent,
  StyledTypography,
} from './MovieInfo.styles';
import { MovieInfo as MovieInfoType } from '../../generated/types';

type Props = {
  movie: MovieInfoType;
  loading: boolean;
  addOrRemoveFromFavorites: () => void;
};

const MovieInfo = ({ movie, addOrRemoveFromFavorites, loading }: Props) => (
  <StyledCard elevation={10}>
    <StyledCardMedia image={movie.backdrop_path || 'no-image.jpg'} src="img" />
    <StyledCardContent>
      <StyledTypography gutterBottom variant="h5">
        {movie.title}
        <Button
          color={movie.isInFavorites ? 'secondary' : 'primary'}
          data-testid="favorites-button"
          disabled={loading}
          onClick={addOrRemoveFromFavorites}
          variant="contained"
        >
          {/* eslint-disable-next-line no-nested-ternary */}
          {loading
            ? 'Saving'
            : movie.isInFavorites
            ? 'Remove from favorites'
            : 'Add to favorites'}
        </Button>
      </StyledTypography>
      <Typography>{movie.overview}</Typography>
      <Divider />
      <Typography>
        <b>Budget:</b> {movie.budget}
      </Typography>
      <Divider />
      <Typography>
        <b>Revenue:</b> {movie.revenue}
      </Typography>
      <Divider />
      <Typography>
        <b>Rating:</b> {movie.vote_average}
      </Typography>
      <Divider />
      <Typography>
        <b>Release Date:</b> {movie.release_date}
      </Typography>
    </StyledCardContent>
  </StyledCard>
);

export default MovieInfo;
