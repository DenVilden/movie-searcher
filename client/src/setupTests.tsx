import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { ThemeProvider } from 'styled-components';
import { theme } from './pages/_app';
import defaultResolvers from './graphql/resolvers';

window.HTMLElement.prototype.scrollIntoView = jest.fn();

type RenderApolloOptions = {
  mocks?: MockedResponse[];
  addTypeName?: boolean;
  cache?: any;
  resolvers?: any;
};

export const renderApollo = (
  element: React.ReactElement,
  { mocks, addTypeName, resolvers, cache }: RenderApolloOptions = {}
) =>
  render(
    <MockedProvider
      addTypename={addTypeName}
      cache={cache}
      mocks={mocks}
      resolvers={resolvers || defaultResolvers}
    >
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MockedProvider>
  );

export * from '@testing-library/react';
