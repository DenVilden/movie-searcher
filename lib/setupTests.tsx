/* eslint-disable import/no-extraneous-dependencies */
import { render } from '@testing-library/preact'
import { MockedProvider, MockedResponse } from '@apollo/client/testing'
import { ThemeProvider, createMuiTheme } from '@material-ui/core/styles'
import '@testing-library/jest-dom/extend-expect'

const theme = createMuiTheme()

type RenderApolloOptions = {
  addTypeName?: boolean
  mocks?: MockedResponse[]
}

export function renderApollo(
  element: React.ReactElement,
  { mocks = [], addTypeName = false }: RenderApolloOptions = {},
) {
  return render(
    <MockedProvider addTypename={addTypeName} mocks={mocks}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MockedProvider>,
  )
}

export { screen, fireEvent } from '@testing-library/preact'
