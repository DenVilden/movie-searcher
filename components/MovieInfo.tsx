import {
  Typography,
  CardContent,
  Card,
  Divider,
  Button,
  Slide,
} from '@material-ui/core';
import styled from '@emotion/styled';
import { useReactiveVar } from '@apollo/client/react';
import Image from 'next/image';
import Head from 'next/head';

import { favoritesVar } from 'apollo/client';
import { Favorite } from 'utils/types';
import { SimilarMovies } from 'apollo/__generated__';
import MoviesBox from './MoviesBox';

const StyledCard = styled(Card)`
  background-color: inherit;
  display: block;
  margin: ${({ theme }) => theme.spacing(11, 1, -8, 1)};

  ${({ theme }) => theme.breakpoints.up('md')} {
    display: flex;
  }
`;

const ImageWrapper = styled.div`
  height: 450px;
  position: relative;
  width: 100%;

  ${({ theme }) => theme.breakpoints.up('md')} {
    width: 40%;
  }
`;

const StyledCardContent = styled(CardContent)`
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  ${({ theme }) => theme.breakpoints.up('md')} {
    width: 60%;
  }
`;

const StyledTypography = styled(Typography)`
  display: flex;
  justify-content: space-between;
`;

interface Props {
  data: {
    backdrop_path?: string | null;
    budget?: string;
    id: number;
    media_type: string;
    number_of_episodes?: number;
    number_of_seasons?: number;
    overview?: string | null;
    poster_path?: string | null;
    release_date: string;
    revenue?: string;
    similar: SimilarMovies;
    title: string;
    vote_average: number;
  };
}

export default function MovieInfoComponent({ data }: Props) {
  const favorites = useReactiveVar(favoritesVar);

  const isInFavorites = favorites.some(favorite => favorite.id === data.id);

  const addOrRemoveFromFavorites = () => {
    let newFavorites: Favorite[];

    if (isInFavorites) {
      newFavorites = favorites.filter(favorite => favorite.id !== data.id);
      favoritesVar(newFavorites);
    } else {
      newFavorites = [
        ...favorites,
        {
          id: data.id,
          media_type: data.media_type,
          poster_path: data.poster_path,
          title: data.title,
        },
      ];
      favoritesVar(newFavorites);
    }

    localStorage.setItem('favorites', JSON.stringify(newFavorites));
  };

  return (
    <Slide direction="up" in>
      <div>
        <Head key="title">
          <title>{data.title}</title>
        </Head>
        <StyledCard elevation={10}>
          <ImageWrapper>
            <Image
              alt={data.title}
              layout="fill"
              objectFit="cover"
              src={data.backdrop_path || '/no-image.jpg'}
            />
          </ImageWrapper>
          <StyledCardContent>
            <StyledTypography gutterBottom variant="h5">
              {data.title}
              <Button
                color={isInFavorites ? 'secondary' : 'primary'}
                onClick={addOrRemoveFromFavorites}
                variant="contained"
              >
                {isInFavorites ? 'Remove from favorites' : 'Add to favorites'}
              </Button>
            </StyledTypography>
            <Typography>{data.overview}</Typography>
            <Divider />
            <Typography>
              {data.media_type === 'movie' ? <b>Budget:</b> : <b>Seasons:</b>}{' '}
              {data.budget || data.number_of_seasons}
            </Typography>
            <Divider />
            <Typography>
              {data.media_type === 'movie' ? <b>Revenue:</b> : <b>Episodes:</b>}{' '}
              {data.revenue || data.number_of_episodes}
            </Typography>
            <Divider />
            <Typography>
              <b>Rating:</b> {data.vote_average}
            </Typography>
            <Divider />
            <Typography>
              <b>Release Date:</b> {data.release_date}
            </Typography>
          </StyledCardContent>
        </StyledCard>
        {!!data.similar.results.length && (
          <MoviesBox movies={data.similar.results} title="Similar Movies" />
        )}
      </div>
    </Slide>
  );
}
