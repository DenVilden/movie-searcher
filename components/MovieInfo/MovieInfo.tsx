import {
  Typography,
  CardContent,
  Card,
  Divider,
  Button,
} from '@material-ui/core';
import styled from 'styled-components';
import Head from 'next/head';
import { useReactiveVar } from '@apollo/client';
import Image from 'next/image';
import { useEffect } from 'react';
import { favoritesVar, GetMovieInfoQuery } from '../../apollo';

const StyledCard = styled(Card)`
  background-color: inherit;
  display: block;
  margin: ${(props) => props.theme.spacing(2)}px;

  ${(props) => props.theme.breakpoints.up('md')} {
    display: flex;
  }
`;

const ImageWrapper = styled.div`
  height: 450px;
  position: relative;
  width: 100%;

  ${(props) => props.theme.breakpoints.up('md')} {
    width: 40%;
  }
`;

const StyledCardContent = styled(CardContent)`
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  ${(props) => props.theme.breakpoints.up('md')} {
    width: 60%;
  }
`;

const StyledTypography = styled(Typography)`
  display: flex;
  justify-content: space-between;
`;

interface Props {
  data: GetMovieInfoQuery;
}

const MovieInfo = ({ data }: Props) => {
  const favorites = useReactiveVar(favoritesVar);

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  const isInFavorites = favorites.some(
    (favorite) => favorite.id === data.movieInfo.id,
  );

  const addOrRemoveFromFavorites = () => {
    if (isInFavorites) {
      favoritesVar(
        favorites.filter((favorite) => favorite.id !== data.movieInfo.id),
      );
    } else {
      favoritesVar([...favorites, data.movieInfo]);
    }
  };

  return (
    <StyledCard elevation={10}>
      <Head>
        <title>{data.movieInfo.title}</title>
      </Head>
      <ImageWrapper>
        <Image
          layout="fill"
          objectFit="cover"
          alt={data.movieInfo.title}
          src={data.movieInfo.backdrop_path || '/no-image.jpg'}
        />
      </ImageWrapper>
      <StyledCardContent>
        <StyledTypography gutterBottom variant="h5">
          {data.movieInfo.title}
          <Button
            color={isInFavorites ? 'secondary' : 'primary'}
            data-testid="favorites-button"
            onClick={addOrRemoveFromFavorites}
            variant="contained"
          >
            {isInFavorites ? 'Remove from favorites' : 'Add to favorites'}
          </Button>
        </StyledTypography>
        <Typography>{data.movieInfo.overview}</Typography>
        <Divider />
        <Typography>
          <b>Budget:</b> {data.movieInfo.budget}
        </Typography>
        <Divider />
        <Typography>
          <b>Revenue:</b> {data.movieInfo.revenue}
        </Typography>
        <Divider />
        <Typography>
          <b>Rating:</b> {data.movieInfo.vote_average}
        </Typography>
        <Divider />
        <Typography>
          <b>Release Date:</b> {data.movieInfo.release_date}
        </Typography>
      </StyledCardContent>
    </StyledCard>
  );
};

export default MovieInfo;
