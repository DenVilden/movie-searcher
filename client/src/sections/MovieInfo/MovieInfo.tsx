import { Typography, Divider, Button } from "@material-ui/core";
import Head from "next/head";
import { useReactiveVar } from "@apollo/client";
import { MovieInfo as MovieInfoType } from "../../graphql";
import {
  StyledCard,
  StyledCardMedia,
  StyledCardContent,
  StyledTypography,
} from "./MovieInfo.styles";
import { favoritesVar } from "../../lib/apollo";

type Props = {
  movie: MovieInfoType;
};

const MovieInfo = ({ movie }: Props) => {
  const favorites = useReactiveVar(favoritesVar);

  const isInFavorites = Object.values(favorites).some(
    (favorite) => favorite.id === movie.id
  );

  const addOrRemoveFromFavorites = () => {
    if (isInFavorites) {
      favoritesVar(
        Object.values(favorites).filter((favorite) => favorite.id !== movie.id)
      );
    } else {
      favoritesVar([...favorites, movie]);
    }
  };

  return (
    <StyledCard elevation={10}>
      <Head>
        <title>{movie.title}</title>
      </Head>
      <StyledCardMedia
        image={movie.backdrop_path || "no-image.jpg"}
        src="img"
      />
      <StyledCardContent>
        <StyledTypography gutterBottom variant="h5">
          {movie.title}
          <Button
            color={isInFavorites ? "secondary" : "primary"}
            data-testid="favorites-button"
            onClick={addOrRemoveFromFavorites}
            variant="contained"
          >
            {isInFavorites ? "Remove from favorites" : "Add to favorites"}
          </Button>
        </StyledTypography>
        <Typography>{movie.overview}</Typography>
        <Divider />
        <Typography>
          <b>Budget:</b> {movie.budget}
        </Typography>
        <Divider />
        <Typography>
          <b>Revenue:</b> {movie.revenue}
        </Typography>
        <Divider />
        <Typography>
          <b>Rating:</b> {movie.vote_average}
        </Typography>
        <Divider />
        <Typography>
          <b>Release Date:</b> {movie.release_date}
        </Typography>
      </StyledCardContent>
    </StyledCard>
  );
};

export default MovieInfo;
