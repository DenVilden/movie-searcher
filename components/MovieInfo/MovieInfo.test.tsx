import MovieInfo from './MovieInfo';
import { renderApollo } from '../../lib/setupTests';

const mocks = {
  movieInfo: {
    id: 1,
    backdrop_path: null,
    poster_path: null,
    title: 'test',
    overview: 'test data',
    budget: '0',
    revenue: '0',
    vote_average: 5,
    release_date: '2020',
    similar: {
      results: [
        {
          id: 1,
          title: 'test',
          release_date: '2020',
          poster_path: null,
        },
      ],
    },
  },
};

describe('movieInfo', () => {
  it('should take a snapshot', () => {
    const { asFragment } = renderApollo(<MovieInfo data={mocks} />);

    const element = asFragment();

    expect(element).toMatchSnapshot();
  });
});
