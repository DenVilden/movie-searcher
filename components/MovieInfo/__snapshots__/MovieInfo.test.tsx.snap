// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movieInfo should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root MovieInfo__StyledCard-a3oxdg-0 inEIfx MuiPaper-elevation10 MuiPaper-rounded"
  >
    <div
      class="MovieInfo__ImageWrapper-a3oxdg-1 iRSvHa"
    >
      <div
        style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
      >
        <img
          alt="test"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
        />
      </div>
    </div>
    <div
      class="MuiCardContent-root MovieInfo__StyledCardContent-a3oxdg-2 dVGSdV"
    >
      <h5
        class="MuiTypography-root MovieInfo__StyledTypography-a3oxdg-3 iLSPhu MuiTypography-h5 MuiTypography-gutterBottom"
      >
        test
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          data-testid="favorites-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add to favorites
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </h5>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        test data
      </p>
      <hr
        class="MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        <b>
          Budget:
        </b>
         0
      </p>
      <hr
        class="MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        <b>
          Revenue:
        </b>
         0
      </p>
      <hr
        class="MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        <b>
          Rating:
        </b>
         5
      </p>
      <hr
        class="MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        <b>
          Release Date:
        </b>
         2020
      </p>
    </div>
  </div>
</DocumentFragment>
`;
